x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/cache/blob/BlobStoreCacheMaintenanceService.java

```diff
@@ -0,0 +1,203 @@
+        final Set<String> paths = new HashSet<>(numberOfShards);
+        for (int shard = 0; shard < numberOfShards; shard++) {
+            paths.add(String.join("/", snapshotUuid, indexUuid, String.valueOf(shard)));
+        }
+        return QueryBuilders.termsQuery("blob.path", paths);
+    }
+
+    private class MaintenanceTask extends AbstractRunnable {
+
+        private final ClusterChangedEvent event;
+
+        MaintenanceTask(ClusterChangedEvent event) {
+            assert event.indicesDeleted().isEmpty() == false;
+            this.event = Objects.requireNonNull(event);
+        }
+
+        @Override
+        protected void doRun() {
+            final ClusterState state = event.state();
+            for (Index deletedIndex : event.indicesDeleted()) {
+                final IndexMetadata indexMetadata = event.previousState().metadata().index(deletedIndex);
+                if (indexMetadata != null) {
+                    final Settings indexSetting = indexMetadata.getSettings();
+                    if (SearchableSnapshotsSettings.isSearchableSnapshotStore(indexSetting)) {
+                        assert state.metadata().hasIndex(deletedIndex) == false;
+
+                        final SnapshotId snapshotId = new SnapshotId(
+                            SNAPSHOT_SNAPSHOT_NAME_SETTING.get(indexSetting),
+                            SNAPSHOT_SNAPSHOT_ID_SETTING.get(indexSetting)
+                        );
+                        final IndexId indexId = new IndexId(
+                            SNAPSHOT_INDEX_NAME_SETTING.get(indexSetting),
+                            SNAPSHOT_INDEX_ID_SETTING.get(indexSetting)
+                        );
+
+                        // we should do nothing if the current cluster state contains another
+                        // searchable snapshot index that uses the same index snapshot
+                        if (hasSearchableSnapshotWith(state, snapshotId, indexId)) {
+                            logger.debug(
+                                "snapshot [{}] of index {} is in use, skipping maintenance of snapshot blob cache entries",
+                                snapshotId,
+                                indexId
+                            );
+                            continue;
+                        }
+
+                        final DeleteByQueryRequest request = new DeleteByQueryRequest(systemIndexName);
+                        request.setQuery(buildDeleteByQuery(indexMetadata.getNumberOfShards(), snapshotId.getUUID(), indexId.getId()));
+                        clientWithOrigin.execute(DeleteByQueryAction.INSTANCE, request, new ActionListener<>() {
+                            @Override
+                            public void onResponse(BulkByScrollResponse response) {
+                                logger.debug(
+                                    "snapshot blob cache maintenance task deleted [{}] documents for snapshot [{}] of index {}",
```

Comment:
nit: should we also mention the deleted mounted index name in this message (and in the failure message)?

Reply:
++